modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -74,48 +168,117 @@ object GlobalL0Service {
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=$ordinal")
+                  .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+              }
+
+          case (Nil, snapshots) =>
+            Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      private def getLatestMajoritySnapshotData(peerIds: NonEmptyList[PeerId]): F[MajoritySnapshotData] = {
+        val maybeData =
+          for {
+            peers <- OptionT(getL0Peers(peerIds))
+            majorityOrdinal <- OptionT(getMajorityOrdinal(peers))
+            majorityHash <- OptionT(getMajorityHash(peers, majorityOrdinal))
+          } yield MajoritySnapshotData(majorityOrdinal, majorityHash)
+
+        maybeData.getOrElseF(NoMajoritySnapshotData.raiseError[F, MajoritySnapshotData])
+      }
+
+      private def getMajorityHash(peers: NonEmptyList[L0Peer], ordinal: SnapshotOrdinal): F[Option[Hash]] =
+        peers.toList
+          .parTraverseN(10)(p =>
+            attemptOption(l0GlobalSnapshotClient.getHash(ordinal).run(p), s"Unable to obtain hash for majority peer $p")
+          )
+          .map(_.flatMap(_.flatten))
+          .flatTap(hashes => logger.debug(s"** Majority Hashes $hashes"))
+          .map(selectMajorityHash)
+
+      private def getPeerInTheMajority(ordinal: SnapshotOrdinal, hash: Hash, peerSet: NonEmptySet[L0Peer]): F[L0Peer] =
+        Random
+          .shuffle(List.from(peerSet.toSortedSet))
+          .tailRecM[F, L0Peer] {
+            case Nil => NoPeerAlignedWithMajority.raiseError[F, Either[List[L0Peer], L0Peer]]
+            case head :: tail =>
+              l0GlobalSnapshotClient
+                .getHash(ordinal)
+                .run(head)
+                .map {
+                  case Some(peerHash) if peerHash === hash => head.asRight[List[L0Peer]]
+                  case _                                   => tail.asLeft[L0Peer]
+                }
+          }
+
+      private def getL0Peers(peerIds: NonEmptyList[PeerId]): F[Option[NonEmptyList[L0Peer]]] =
+        peerIds
+          .traverse(globalL0ClusterStorage.getPeer)
+          .map(_.toList.flatten.toNel)
+
+      private def getMajorityOrdinal(peers: NonEmptyList[L0Peer]): F[Option[SnapshotOrdinal]] =
+        peers.toList
+          .parTraverseN(10)(p =>
+            attemptOption(l0GlobalSnapshotClient.getLatestOrdinal.run(p), s"Unable to retrieve ordinal for majority peer $p")
+          )
+          .map(_.flatten)
+          .map(pickMajority[List, SnapshotOrdinal])
+
+      private def selectMajorityHash(hashes: List[Hash]): Option[Hash] = {
+        val minCount = max(minimumMajorityPeerIds, ceil(0.5 * (1 + hashes.size)).toInt)
```

Comment:
I think you are calculating the `minCount` wrongly here - e.i. based on the amount of hashes rather than the original amount of majority peers. In other places in the code you are flattening the results when fetching hashes from a node or ordinal or fetching peer from the clusterStorage which I would think about as well because some of these conditions may be a reason to short circuit already. But as long as you will have the correct majority counting logic these flattening should not have a bad effect, the thing is the logic is wrong. You can end with two hashes from e.g. 10 initial majority peers (some not present in the storage, some don't have the hash for ordinal and so more and more of them are kind of discarded) in that function and you would again calculate the outcome based on the minority of nodes. If we have 10 majority peers we can't determine majority unless majority of them agrees. So for the situation you got 6 responds out of 10, then they all need to agree. You actually can also have it work for more fragmented clusters - if you have all the responses from 10 nodes and the majority is 4 -> A, 3 -> B, 3 -> C then you can also make a decision that A is the majority hash. Not sure if we want to go in that direction though, we may but there can be corner cases there - I will try to elaborate tomorrow.

Reply:
@TheMMaciek
Re. using no. of peers to calculate minimum rather than no. of retrieved hashes, is it likely that some of the majority peers become unresponsive for an indefinite amount of time?

Re. the 4,3,3 counts, if we accept 4 as the majority then that doesn't conform to the > 50% formula.

Reply:
Well the nodes becoming unresponsive - that is hard to estimate, if too many fail to respond then your process will simply get stuck. But I think it's essential to get it stuck and not make a wrong decision.

The 4,3,3 matter: yeah you are right it doesn't conform to the >50% formula but it conforms to the deterministic decision property we need - 4 is still a majority but we can only state that in the situation described when all 10 nodes send their responses. So I proposed it only because such situation also gives us determinism.

Reply:
@TheMMaciek @sdavidp 
> But I think it's essential to get it stuck and not make a wrong decision.

IMO it's essential to make deterministic decision. you don't know it's right or wrong, but it must be deterministic (which means that future responses does not influence or won't change the decision)

Reply:
@marcinwadon @TheMMaciek @ryle-goehausen 

Right now the formula we discussed for selecting the majority is only being applied to find the majority hash: determining the majority ordinal is only calling _pickMajority_

Should we use the same formula for finding majority ordinal?